tests/test_publish_dafsa.py

```diff
@@ -0,0 +1,148 @@
+            adding_headers={"Content-Type": "application/json"},
+            json='{"data": {"password": "pAsSwErD"}}',
+        )
+        client = Client(
+            server_url=server,
+            auth=("arpit73", "pAsSwErD"),
+            bucket=BUCKET_ID,
+            collection=COLLECTION_ID,
+        )
+        self.assertEqual(
+            client.get_endpoint("record", id=RECORD_ID),
+            "/buckets/dafsa-bucket/collections/public-suffix-list/records/tld-dafsa",
+        )
+
+        responses.add(
+            responses.POST,
+            attachment_uri,
+            adding_headers={"Content-Type": "multipart/form-data"},
+            files={"attachment": ("dafsa.bin", b"some binary data")},
+            json='{"commit_hash": "abc"}',
+        )
+        responses.add(
+            responses.PATCH,
+            record_uri,
+            adding_headers={"Content-Type": "application/json"},
+            json='{"data": {"status": "to-review"}}',
+        )
+
+        with tempfile.TemporaryDirectory() as tmp:
+            with open(f"{tmp}/dafsa.bin", "wb") as f:
+                f.write(b"some binary data")
+            remote_settings_publish(client, "abc", f"{tmp}/dafsa.bin")
+
+
+class TestPublishDafsa(unittest.TestCase):
+    event = {"server": "https://kinto.dev.mozaws.net/v1", "auth": "arpit73:pAsSwErD"}
+    record_uri = f"{{{event.get('server')}/buckets/main-workspace/collections/public-suffix-list/records/tld-dafsa}}"  # noqa
+
+    def setUp(self):
+        mocked = mock.patch("commands.publish_dafsa.prepare_dafsa")
+        self.addCleanup(mocked.stop)
+        self.mocked_prepare = mocked.start()
+
+        mocked = mock.patch("commands.publish_dafsa.remote_settings_publish")
+        self.addCleanup(mocked.stop)
+        self.mocked_publish = mocked.start()
+
+    @responses.activate
+    def test_prepare_and_publish_are_not_called_when_hash_matches(self):
+        responses.add(responses.GET, COMMIT_HASH_URL, json=[{"sha": "abc"}])
+        responses.add(
+            responses.GET, self.record_uri, json={"data": {"commit-hash": "abc"}}
+        )
+
+        publish_dafsa(self.event, context=None)
+
+        self.assertFalse(self.mocked_prepare.called)
+        self.assertFalse(self.mocked_publish.called)
+
+    @responses.activate
+    def test_KintoException_raised_when_fetching_failed(self):
+        responses.add(responses.GET, COMMIT_HASH_URL + "c", json=[{"sha": "abc"}])
+        responses.add(
+            responses.PUT,
+            f"{{{self.event.get('server')}/accounts/arpit73}}",
+            adding_headers={"Content-Type": "application/json"},
+            json='{"data": {"password": "pAsSwErD"}}',
+        )
+        responses.add(
+            responses.GET, self.record_uri, json={"data": {"commit-hash": "abc"}}
+        )
+
+        with self.assertRaises(KintoException) as e:
+            publish_dafsa(self.event, context=None)
+            self.assertEqual(e.response, None) or self.assertNotEqual(
```

Comment:
No, here you should know what `e.response`, no `or` is necessary

Reply:
We raise on two conditions so that's why I used the `or`
```python
except KintoException as e:
        if e.response is None or e.response.status_code != 404:
            raise
```

Reply:
But that's not how tests work.

The code raises in two different situations, indeed.

Your test creates a situation, and assert something about what happened.

If you want to assert what happened in two different situations, then you need two tests. Each of them will have its own assertion. 

Now. Back to this particular case. Testing about the 404 is enough. We only care about the 404, and we have this `if e.response is None or` just to avoid the risk to have an `AttributeError` in case `e.response` is not set.